tests/test_simulate.py

```diff
@@ -0,0 +1,230 @@
+import jax.numpy as jnp
+import pytest
+from jax import random
+from lcm.entry_point import (
+    create_compute_conditional_continuation_policy,
+    get_lcm_function,
+    get_next_state_function,
+)
+from lcm.example_models import PHELPS_DEATON, PHELPS_DEATON_WITH_FILTERS
+from lcm.model_functions import get_utility_and_feasibility_function
+from lcm.process_model import process_model
+from lcm.simulate import (
+    _retrieve_non_sparse_choices,
+    create_choice_segments,
+    create_data_scs,
+    dict_product,
+    filter_ccv_policy,
+    simulate,
+)
+from lcm.state_space import create_state_choice_space
+from numpy.testing import assert_array_equal
+
+# ======================================================================================
+# Test simulate using raw inputs
+# ======================================================================================
+
+
+@pytest.fixture()
+def simulate_inputs():
+    user_model = {**PHELPS_DEATON, "n_periods": 1}
+    model = process_model(user_model)
+
+    _, space_info, _, _ = create_state_choice_space(
+        model=model,
+        period=0,
+        jit_filter=False,
+    )
+
+    next_state = get_next_state_function(model)
+    u_and_f = get_utility_and_feasibility_function(
+        model=model,
+        space_info=space_info,
+        data_name="vf_arr",
+        interpolation_options={},
+        is_last_period=True,
+    )
+
+    compute_ccv_policy_functions = model.n_periods * [
+        create_compute_conditional_continuation_policy(
+            utility_and_feasibility=u_and_f,
+            continuous_choice_variables=["consumption"],
+        ),
+    ]
+
+    return {
+        "state_indexers": [{}],
+        "continuous_choice_grids": [{"consumption": jnp.linspace(1, 100, num=11)}],
+        "compute_ccv_policy_functions": compute_ccv_policy_functions,
+        "model": model,
+        "next_state": next_state,
+    }
+
+
+def test_simulate_using_raw_inputs(simulate_inputs):
+    params = {
+        "beta": 1.0,
+        "utility": {"delta": 1.0},
+        "next_wealth": {
+            "interest_rate": 0.05,
+            "wage": 1.0,
+        },
+    }
+
+    got = simulate(
+        params=params,
+        vf_arr_list=[None],
+        initial_states={"wealth": jnp.array([1.0, 50.5])},
+        **simulate_inputs,
+    )
+
+    choices = got[0]["choices"]
+
+    assert_array_equal(choices["retirement"], 1)
```

Comment:
Why exactly do we expect that the agent is not working in the last period? If I am not mistaken she could consume more than her current wealth which might be optimal in some cases.

Reply:
I rewrote the constraint, so this should be clearer. We have the borrowing constraint:

```python
consumption <= wealth
```

and the wealth transition

```python
next_wealth = (1 + interest_rate) (wealth - consumption) + wage * working
```

If you work in period t, your wealth in period t+1 increases by `wage`. But in your last period, you do not care about the wealth in your next period. One could discuss whether this timing assumption is sensible, but it seems like this is how it is formulated by Iskhakov et al. (2017).

Reply:
Ah, then it makes sense.

I thought we were using this borrowing constraint:

```

def next_wealth_constraint(next_wealth):
    return next_wealth >= 0
    
```

I don't have a good intution on whether we would like to use the `next_wealth_constraint` or `consumption_constraint`. We can discuss it later.